dev/devicelab/lib/framework/browser.dart

```diff
@@ -136,3 +209,256 @@ String _findSystemChromeExecutable() {
+        })
+        .single;
+      final int tabPid = processLabel.pid;
+      events = events.where((BlinkTraceEvent element) => element.pid == tabPid).toList();
+
+      // Extract frame data.
+      final List<BlinkFrame> frames = <BlinkFrame>[];
+      int skipCount = 0;
+      BlinkFrame frame = BlinkFrame();
+      for (final BlinkTraceEvent event in events) {
+        if (event.ph == 'X' && event.name == 'WebViewImpl::beginFrame') {
+          frame.beginFrame = event;
+        } else if (event.ph == 'X' && event.name == 'WebViewImpl::updateAllLifecyclePhases') {
+          frame.updateAllLifecyclePhases = event;
+          if (frame.endMeasuredFrame != null) {
+            frames.add(frame);
+          } else {
+            skipCount += 1;
+          }
+          frame = BlinkFrame();
+        } else if (event.ph == 'b' && event.name == 'measured_frame') {
+          frame.beginMeasuredFrame = event;
+        } else if (event.ph == 'e' && event.name == 'measured_frame') {
+          frame.endMeasuredFrame = event;
+        }
+      }
+
+      print('Extracted ${frames.length} measured frames.');
+      print('Skipped $skipCount non-measured frames.');
+
+      if (frames.isEmpty) {
+        // The benchmark is not measuring frames.
+        return null;
+      }
+
+      // Compute averages and summarize.
+      return BlinkTraceSummary._(
+        averageBeginFrameTime: _computeAverageDuration(frames.map((BlinkFrame frame) => frame.beginFrame).toList()),
+        averageUpdateLifecyclePhasesTime: _computeAverageDuration(frames.map((BlinkFrame frame) => frame.updateAllLifecyclePhases).toList()),
+      );
+    } catch (_, __) {
+      final io.File traceFile = io.File('./chrome-trace.json');
+      io.stderr.writeln('Failed to interpret the Chrome trace contents. The trace was saved in ${traceFile.path}');
+      traceFile.writeAsStringSync(const JsonEncoder.withIndent('  ').convert(traceJson));
+      rethrow;
+    }
+  }
+
+  /// The average duration of "WebViewImpl::beginFrame" events.
+  ///
+  /// This event contains all of scripting time of an animation frame, plus an
+  /// unknown small amount of work browser does before and after scripting.
+  final Duration averageBeginFrameTime;
+
+  /// The average duration of "WebViewImpl::updateAllLifecyclePhases" events.
+  ///
+  /// This event contains style, layout, painting, and compositor computations,
+  /// which are not included in the scripting time. This event does not
+  /// include GPU time, which happens on a separate thread.
+  final Duration averageUpdateLifecyclePhasesTime;
+
+  /// The average sum of [averageBeginFrameTime] and
+  /// [averageUpdateLifecyclePhasesTime].
+  ///
+  /// This value contains the vast majority of work the UI thread performs in
+  /// any given animation frame.
+  final Duration averageTotalUIFrameTime;
+
+  @override
+  String toString() => '$BlinkTraceSummary('
```

Comment:
I see this pattern a lot in Dart code (using `$ClassName` instead of plain `ClassName`). Is there a reason for it? The only reason I can think of is that the string doesn't get out of date if the class is renamed (the analyzer would complain that `$OldClassName` isn't defined). Is that it?

Reply:
... and in comments.